src/RA_StringUtils.h

```diff
@@ -25,65 +25,486 @@ _NODISCARD std::string Narrow(_In_ const std::string& wstr);
+{
+    return ra::Widen(value);
+}
+
+template<> _NODISCARD inline const std::wstring ToWString(_In_ char* value)
+{
+    return ra::Widen(value);
+}
+
+template<> _NODISCARD inline const std::wstring ToWString(_In_ const char value)
+{
+    return std::wstring(1, value);
+}
+
+// ----- string building -----
+
+class StringBuilder
+{
+public:
+    StringBuilder(bool bPrepareWide = false)
+        : m_bPrepareWide(bPrepareWide)
+    {
+    }
+
+    template<typename T>
+    void Append(T arg)
+    {
+        PendingString pPending;
+        if (m_bPrepareWide)
+        {
+            pPending.WString = ra::ToWString(arg);
+            pPending.DataType = PendingString::Type::WString;
+        }
+        else
+        {
+            pPending.String = ra::ToString(arg);
+            pPending.DataType = PendingString::Type::String;
+        }
+        m_vPending.emplace_back(std::move(pPending));
+    }
+
+    template<>
+    void Append(const std::string& arg)
+    {
+        PendingString pPending;
+        pPending.Ref.String = &arg;
+        pPending.DataType = PendingString::Type::StringRef;
+        m_vPending.emplace_back(std::move(pPending));
+    }
+
+    template<>
+    void Append(const std::wstring& arg)
+    {
+        PendingString pPending;
+        pPending.Ref.WString = &arg;
+        pPending.DataType = PendingString::Type::WStringRef;
+        m_vPending.emplace_back(std::move(pPending));
+    }
+
+    template<>
+    void Append(const char* arg)
+    {
+        AppendSubString(arg, strlen(arg));
+    }
+
+    template<>
+    void Append(const wchar_t* arg)
+    {
+        AppendSubString(arg, wcslen(arg));
+    }
+
+    void AppendSubString(const char* pStart, size_t nLength)
+    {
+        if (nLength == 0)
+            return;
 
-    va_list pArgs;
-    va_start(pArgs, sFormat);
-    int nNeeded = 0;
-    if constexpr(std::is_same_v<CharT, char>)
+        PendingString pPending;
+        pPending.Ref.Char.Pointer = pStart;
+        pPending.Ref.Char.Length = nLength;
+        pPending.DataType = PendingString::Type::CharRef;
+        m_vPending.emplace_back(std::move(pPending));
+    }
+
+    void AppendSubString(const wchar_t* pStart, size_t nLength)
     {
-        assert(std::string_view{ sFormat }.find("%n") == std::string_view::npos);
-        assert(sFormatted.capacity() > 0U && (sFormatted.capacity() < RSIZE_MAX/sizeof(char)));
-        nNeeded = std::vsnprintf(sFormatted.data(), sFormatted.capacity(), sFormat, pArgs);
+        if (nLength == 0)
+            return;
+
+        PendingString pPending;
+        pPending.Ref.WChar.Pointer = pStart;
+        pPending.Ref.WChar.Length = nLength;
+        pPending.DataType = PendingString::Type::WCharRef;
+        m_vPending.emplace_back(std::move(pPending));
     }
-    else if constexpr(std::is_same_v<CharT, wchar_t>)
+
+    template<typename T, typename... Ts>
+    void Append(T arg, Ts... args)
```

Comment:
`arg` _should_ probably be passed by `const reference`, `args` definitely needs be passed by `rvalue reference`.

Reply:
done